### PR TITLE
[Core] RatesProvider

### PR DESCRIPTION
**Pull request by flypaper0**

### CODE REVIEW

File: VultisigApp/VultisigApp/Services/CryptoPriceService.swift [hunk 1]
```diff
@@ -1,120 +1,124 @@
...
+            currencies: currencies
+        )
+
+        let (data, _) = try await URLSession.shared.data(from: url)
+        let response = try JSONDecoder().decode([String: [String: Double]].self, from: data)
+
+        try await RateProvider.shared.save(rates: mapRates(response: response))
+    }
+
+    func fetchPrices(contracts: [String], chain: Chain) async throws {
```
Comment (johnnyluo) [inline VultisigApp/VultisigApp/Services/CryptoPriceService.swift:91]: Previously , we cache the price for a while , with this new implementation , rates are stored in database , and how often do we refresh the rates?

File: VultisigApp/VultisigApp/Views/Components/VaultDetailBalanceContent.swift [hunk 2]
```diff
@@ -8,8 +8,8 @@
 import SwiftUI
 
 struct VaultDetailBalanceContent: View {
-    let vault: Vault
-    
+    @ObservedObject var vault: Vault
```
Comment (johnnyluo) [inline VultisigApp/VultisigApp/Views/Components/VaultDetailBalanceContent.swift:11]: why we need to have vault as an ObservableObject? will it be changed ?

**Comment by johnnyluo** [comment]
Good work @flypaper0  , it is a pretty big PR , let's do more tests.

File: VultisigApp/VultisigApp/Services/CryptoPriceService.swift (hunk 1 above)
Comment (flypaper0) [reply VultisigApp/VultisigApp/Services/CryptoPriceService.swift:91]: We're refreshing rates in same places where we refreshing balances. Check BalanceService
If we fetch balances for whole value - fetching all rates
For single coin - fetching rate for this coin only

File: VultisigApp/VultisigApp/Views/Components/VaultDetailBalanceContent.swift (hunk 2 above)
Comment (flypaper0) [reply VultisigApp/VultisigApp/Views/Components/VaultDetailBalanceContent.swift:11]: As we're not storing rate inside coin anymore, we should observe vault to track changes. 
I'm trigger vault update with `vault.objectWillChange.send()` in CryptoPriceService

File: VultisigApp/VultisigApp/Model/Coin.swift [hunk 3]
```diff
@@ -172,11 +173,11 @@ class Coin: ObservableObject, Codable, Hashable {
     
     func fiat(value: BigInt) -> Decimal {
         let decimal = decimal(for: value)
-        return decimal * Decimal(priceRate)
+        return RateProvider.shared.fiatBalance(value: decimal, coin: self)
```
Comment (enriquesouza) [inline VultisigApp/VultisigApp/Model/Coin.swift:176]: I would declare the RateProvider as a var, so we don't have to change everywhere if we need to change it.

**Review by enriquesouza — APPROVED** [review]

**Pull request merged**